libcnb/src/layer/struct_api/mod.rs

```diff
@@ -0,0 +1,221 @@
+    /// See: `inspect_restored` in [`CachedLayerDefinition`].
+    Inspect { cause: IAC },
+}
+
+/// A value-to-value conversion for layer actions.
+///
+/// Similar to [`Into`], but specialized. Allowing it to also be implemented for
+/// values in the standard library such as [`Result`].
+///
+/// Implement this trait if you want to use your own types as actions.
+///
+/// libcnb ships with generic implementations for the majority of the use-cases:
+/// - Using [`InspectRestoredAction`] or [`InvalidMetadataAction`] directly.
+/// - Using [`InspectRestoredAction`] or [`InvalidMetadataAction`] directly, wrapped in a Result.
+/// - Using [`InspectRestoredAction`] or [`InvalidMetadataAction`] with a cause value in a tuple.
+/// - Using [`InspectRestoredAction`] or [`InvalidMetadataAction`] with a cause value in a tuple, wrapped in a Result.
+pub trait IntoAction<T, C, E> {
+    fn into_action(self) -> Result<(T, C), E>;
+}
+
+// Allows to use the layer actions directly.
+impl<T, E> IntoAction<T, (), E> for T {
+    fn into_action(self) -> Result<(T, ()), E> {
+        Ok((self, ()))
+    }
+}
+
+//  Allows to use the layer actions directly wrapped in a Result.
+impl<T, E> IntoAction<T, (), E> for Result<T, E> {
+    fn into_action(self) -> Result<(T, ()), E> {
+        self.map(|value| (value, ()))
+    }
+}
+
+// Allows to use the layer actions directly with a cause as a tuple.
+impl<T, C, E> IntoAction<T, C, E> for (T, C) {
+    fn into_action(self) -> Result<(T, C), E> {
+        Ok(self)
+    }
+}
+
+// Allows to use the layer actions directly with a cause as a tuple wrapped in a Result.
+impl<T, C, E> IntoAction<T, C, E> for Result<(T, C), E> {
+    fn into_action(self) -> Result<(T, C), E> {
+        self
+    }
+}
+
+/// A reference to an existing layer on disk.
+///
+/// Provides functions to modify the layer such as replacing its metadata, environment, SBOMs or
+/// exec.d programs.
+///
+/// To obtain a such a reference, use [`BuildContext::cached_layer`] or [`BuildContext::uncached_layer`].
+pub struct LayerRef<B, MAC, IAC>
+where
+    B: Buildpack + ?Sized,
+{
+    name: LayerName,
+    // Technically not part of the layer itself. However, the functions that modify the layer
+    // will need a reference to the layers directory as they will also modify files outside the
+    // actual layer directory. To make LayerRef nice to use, we bite the bullet and include
+    // the layers_dir here.
+    layers_dir: PathBuf,
+    buildpack: PhantomData<B>,
+    pub state: LayerState<MAC, IAC>,
+}
+
+impl<B, MAC, IAC> LayerRef<B, MAC, IAC>
+where
+    B: Buildpack,
+{
+    /// Returns the path to the layer on disk.
+    pub fn path(&self) -> PathBuf {
+        self.layers_dir.join(self.name.as_str())
+    }
+
+    /// Replaces the existing layer metadata with a new value.
+    ///
+    /// The new value does not have to be of the same type as the existing metadata.
+    pub fn replace_metadata<M>(&self, metadata: M) -> crate::Result<(), B::Error>
```

Comment:
It feels a little strange to have functions named `replace_*` (such as `replace_metadata`, `replace_env`, `replace_sboms` etc) when the layer could be new and its not replacing anything?

What do you think about either no prefix at all, or something like `set_` as a prefix?

eg:

```rust
my_layer_ref.metadata(...);
my_layer_ref.env(...);
my_layer_ref.sboms(...);
my_layer_ref.exec_d_programs(...);
```

Or:

```rust
my_layer_ref.set_metadata(...);
my_layer_ref.set_env(...);
my_layer_ref.set_sboms(...);
my_layer_ref.set_exec_d_programs(...);
```

Reply:
Slack call outcome: we decided to use a `write_` prefix.

Reply:
[06d4668](https://github.com/heroku/libcnb.rs/pull/814/commits/06d466876f9140a27e2eef25cd79726eb4e01223)